pymodbus/server/async_io.py

```diff
@@ -173,17 +175,28 @@ def execute(self, request, *addr):
         if self.server.request_tracer:
             self.server.request_tracer(request, *addr)
 
+        task = self.loop.create_task(self._async_execute(request, *addr))
+        self.request_tasks.add(task)
+        task.add_done_callback(self.request_tasks.discard)
+
+    async def _async_execute(self, request, *addr):
         broadcast = False
+        response = None
         try:
             if self.server.broadcast_enable and not request.slave_id:
                 broadcast = True
                 # if broadcasting then execute on all slave contexts,
                 # note response will be ignored
                 for slave_id in self.server.context.slaves():
                     response = request.execute(self.server.context[slave_id])
+                    if asyncio.iscoroutine(response):
+                        response = await response
             else:
                 context = self.server.context[request.slave_id]
                 response = request.execute(context)
+                if asyncio.iscoroutine(response):
```

Comment:
If you move the "if" outside the "else:" you do not need to duplicate the test.

Remark: the "for slave_id" is not the best programming, because it can only be there once.

Reply:
Ah, good point, I'll move it outside of else block.

Reply:
Only problem with it being outside, is that with broadcast it doesn't raise exception in other than last request. But as this being temporary code, I think it is acceptable transition phase?

Reply:
I left these in place, mainly so that it still raises exceptions from execute in broadcast when it happens and it would end up leaving other execute calls to be not awaited.